Issue 501 os to pathlib

finozzifa opened the pull request:
# Closes # (if applicable).
Issue #501 

## Changes proposed in this Pull Request
The PR aims at replacing the os module with the pathlib module

## Checklist

- [x] I consent to the release of this PR's code under the AGPLv3 license and non-code contributions under CC0-1.0 and CC-BY-4.0.
- [x] I tested my contribution locally and it seems to work fine.
- [ ] Code and workflow changes are sufficiently documented.
- [ ] Newly introduced dependencies are added to `envs/environment.yaml` and `doc/requirements.txt`.
- [ ] Changes in configuration options are added in all of `config.default.yaml` and `config.tutorial.yaml`.
- [ ] Add a test config or line additions to `test/` (note tests are changing the config.tutorial.yaml)
- [ ] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [x] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

finozzifa commented:
Hi all,
just as a side-note, every change performed has been tested locally in ad-hoc python script, where I made sure that the result obtained with the os module was the same as the one I got with the pathlib module

pz-max commented on test/test_helpers.py:
```diff
@@ -0,0 +1,90 @@
+path_cwd = str(pathlib.Path.cwd())
+
+
+def test_get_abs_path():
```
A nice test would be to check if the `os` thing you are replacing is equal to the pathlib solution

finozzifa replied:
done! I have added extra assert statements in test/test_helpers.py :D

pz-max commented on scripts/_helpers.py:
```diff
@@ -821,3 +818,68 @@ def get_last_commit_message(path):
+    """
+    It creates recursively the directory and its leaf directories.
+    """
+    pathlib.Path(path).mkdir(parents=True, exist_ok=True)
```
For consistency, just tagging that abstractions are missing for: 

- `str(pathlib.Path.cwd())`

- `str(pathlib.Path(str(config["destination"])).relative_to("."))` 

- `pathlib.Path(base_folder, "configs").exists()`

Maybe also a few more.

pz-max reviewed (commented):
Really a PR that goes through everything. Just added a comment on 1) abstraction consistency and a 2) tests that makes sure that the os to pathlib swap is faultless. On another note, I am not sure if the abstraction layer is super helpful as the pathlib library is quite readable.

This is just a comment. Davide/Katia have for sure more perspectives on this.

ekatef commented:
> Really a PR that goes through everything. Just added a comment on 1) abstraction consistency and a 2) tests that makes sure that the os to pathlib swap is faultless. On another note, I am not sure if the abstraction layer is super helpful as the pathlib library is quite readable.
> 
> This is just a comment. Davide/Katia have for sure more perspectives on this.

Thanks a lot for reviewing, @pz-max! 

And thank you @finozzifa for the great work :) As discussed, our `"good first issues"` are quite labor-intensive, but you manage it graciously. Also, thanks a lot for the PRs you have completed along the way! From our side, I fear that reviewing can take some time. Apologies for that, and please keep in mind that it's not your fault in any way.

Regarding testing, @finozzifa has mentioned that every change has been tested locally to ensure that the workflow output remained the same. [Windows CI currently crashes due to troubles with GADM server: `Exception: GADM server is down at https://geodata.ucdavis.edu/gadm/gadm4.1/gpkg/gadm41_NGA.gpkg. Data needed for building shapes can't be extracted.`] But I agree that it would be really handy to have some testing output to be sure the introduced changes work, even for some edge cases. @finozzifa do you see any possibility to share the testing infrastructure you have developed? Like creating and pushing a dedicated branch for testing? That is not your word in which I have doubts, rather our ability to detect potential troubles with just an intense look into the code... 🙃

Regarding the consistency, I wonder if a string representation of `Path` object does make sense (e.g. `str(pathlib.Path.cwd())`), as it basically gives the result as `os` methods. [The idea taken from @davide-f  :) ] May it be a better approach to revert the changes, if such a transformation is needed? That would also allow to keep the PR lighter...

@finozzifa @davide-f @pz-max what are your feeling about all that? 🙂

finozzifa commented:
Hi all,

thanks for your interest in my PR. I really enjoyed working on it :)

-) @pz-max:  I created separate methods that enclose pathlib invocations to enable an easier unit testing and to make the code more maintainable. Now, if we need to change one line in (say) method _get_path_, we just need to change one file, rather than the entire **scripts** folder

-) @pz-max and @ekatef: I have created dedicated unit tests, in the _test_helpers.py_ file

-)@pz-max and @ekatef: The unit tests have been created as follows: 1) I create the method with the os module , 2) I build the unit test, 3) I re-write the method with pathlib, 4) I make sure that the corresponding unit test still works. 

-) Indeed @pz-max , I have not created methods for all invocations of pathlib. My goal would be do that and so having the _import pathlib_ just in the __helpers.py_ file. I am happy to do it in this issue, or create a new one and work on it after this PR is merged.

Next steps:
-) I will investigate further why the pipeline failed (thanks @ekatef)
-) I will modify the unit tests, adding a further assert statement where I compare the output with the corresponding os invocation I replaced

FabianHofmann reviewed (commented):
@finozzifa I am having just a very short look at this. I think the idea to move to pathlib Path is great. I also think there is a potential for even leaner code. 


In general, I don't think it is a good idea to introduce abstracted functions with function names nearly as long as the content. For example 

```
def get_posixpath(*args):
    """
    It returns a new PosixPath object.
    """
    return pathlib.Path(*args)

```

should definitely be avoided. Why? Python users (who may well be experienced in Python) don't know this function when they see their usage in the scripts. So on the first sight, they won't know things like

```
dirname = get_posixpath("path/to/my/dir")
```

but they would know operations like

```
dirname = Path("path/to/my/dir")
```

Another downside is that you lose all the doc-strings, argument refs (and potential meta data for static type checking) compared to calling `Path`. 

So, whenever there is very basic, short python code, I would propose to not abstracting it. 


Therefore, I would try to always work with the explicit Path objects in the scripts (Path super powerful!) and use operators like `.parent` or `/` . For example:

- Operations like: 
```
GADM_inputfile_gpkg = get_path(
        str(pathlib.Path.cwd()), "data", "gadm", GADM_filename, GADM_filename + ".gpkg")
```

would be 

```
GADM_inputfile_gpkg = Path.cwd() / "data/gadm" / GADM_filename / GADM_filename + ".gpkg"
```

this is much more readable and everyone knows whats happening. 

This would probably remove the need for most new defined get_ functions and avoid to cycle between `str` and `Path` most of the times.

For the often used `os.chdir(get_dirname_abs_path(__file__))`, that could perhaps be combined with a info, since this might otherwise confuse users?


```
import os
from pathlib import Path

def change_to_script_dir(script_path):
    """
    Change the current working directory to the directory containing the given script.
    
    Parameters:
    script_path (str or Path): The path to the script file.
    """
    # Ensure script_path is a Path object
    script_path = Path(script_path)
    
    # Get the directory containing the script
    script_directory = script_path.parent
    
    # Change the current working directory to the script directory
    os.chdir(script_directory)
    
    # Optionally, print the new working directory
    # potentially use logger.info instead
    print("Changed directory to: ", Path.cwd())

```
call it with `change_to_script_dir(__file__)`

These are just my thoughts, nothing mandatory. Hope that helps.

finozzifa commented:
Hi @FabianHofmann thanks for your comment :)

I believe that I follow a different coding style. Whenever I see a line of code which repeats itself several times, I try (where possible) to create methods that implement the line of code and that can be separately be tested. This is because I want to ensure consistency and make future changes easier. In fact, instead of tracing throughout the code each invocation of (say) _os.path.basename()_, I just need to modify once the method where such line is implemented.

I am not sure why code readability would suffer following this approach, given the fact that people would just need to trace the method in the __helpers.py_ to check what it does.

I believe that in this way, code reviews, code merges and further code expansions will be easier, as we can rely on proof-sound methods that we have unit-tested.

Having said that, I am the last one who join the project. You guys are the code experts for PyPSA so I am happy to follow your style :)

FabianHofmann commented:
I think that is super reasonable. It makes most sense to avoid code duplication. Just felt that some of the functions are not "abstract" enough. Perhaps a common ground is trying to avoid the back and forth between Path and string and trying to stick to Path wherever possible? Btw, f-strings nicely process `Path` objects, like 


```
p  = Path("path/to/file")
print(f"This is the path: {p}")
```

But I said, you have the full power :)

finozzifa commented:
Hi @FabianHofmann,

thanks again for your comment and your contribution :) (which I really appreciate)

I added the casting to string of the PosixPath object because otherwise some parts of code are failing. 

Leveraging on the possibilities offered by the abstract method _get_path_, I removed such casting and restarted the pipelines. As you see they are unfortunately failing. One error that I have found is:

_ERROR:_helpers:An error happened in module '/home/runner/work/pypsa-earth/pypsa-earth/scripts/build_natura_raster.py', function '<listcomp>': argument of type 'PosixPath' is not iterable_

It seems therefore that somewhere the casting is indeed needed :(

FabianHofmann commented:
Sorry for being late here. I am wondering where this happens. Could you point me to the exact spot? The string object is iterable but in case of a path it would not make sense ("path/to" would be divided into "p", "a", "t", ...).

finozzifa commented:
Hi @FabianHofmann, 
thanks a lot :)
I believe that my previous comment was a bit misleading. 
I just wanted to point out that apparently there are some parts of the code where PosixPath objects cannot be used and so their string casting is necessary.
Anyway, I once again remove the string casting from _get_path_ and I am letting the pipelines fail again so that I can perform further investigations.

finozzifa commented:
Hi all,

please find below a recap of the changes I performed in this PR.

I tried my best to include all of your suggestions and recommendations. If I missed some, I apologize and I will include it as soon as possible.

**### Recap of the changes**

1) The os module has been replaced with the pathlib module. There are still a few calls in _build_natura_raster.py_ (this is because there is no pathlib alternative for os.walk) and in the __helpers.py_

2) As mentioned before, I have introduced an abstraction of the path methods. Such methods are available in the __helpers.py_ script. Whenever I see a line of code which repeats itself several times, I try (where possible) to create methods that implement such line of code and that can be separately tested. This is because I want to ensure consistency and make future changes easier. In fact, instead of tracing throughout the code each invocation of (say) os.path.basename(), I just need to modify once the method where such line is implemented. I believe that in this way, code reviews, code merges and further code expansions will be easier, as we can rely on soundproof methods that we have unit-tested.

3) I have introduced dedicated unit tests with pytest for the path methods I coded in the _helpers.py. As of now, these unit tests are not executed as part of the CI/CD pipelines. We could include their execution in the CI/CD in a dedicated issue

**### Possible spin-offs**

1) It would nice to push these modifications a bit further, so that the dependency on the _os_ and _pathlib_ modules is explicit just in the __helpers.py_ script.

2) It would nice to include the execution of the unit tests in the CI/CD pipeline, as already done for the framework repository

I thank you in advance for your comments (and patience) in reviewing this PR :)

Fabrizio
